src/dune_rules/resolve.ml

```diff
@@ -1,34 +1,82 @@
 open Stdune
 open Dune_engine
-include Or_exn
 
-let of_result t = t
+type error =
+  { exn : exn
+  ; stack_frames : Memo.Stack_frame.Info.t Lazy.t list
+  }
+
+type 'a t = ('a, error) result
+
+let return x = Ok x
+
+let bind = Result.bind
+
+include Monad.Make (struct
+  type nonrec 'a t = 'a t
+
+  let return = return
+
+  let bind = bind
+end)
+
+let error_equal { exn; stack_frames } b =
+  Exn.equal exn b.exn
+  && List.equal
+       (fun (lazy a) (lazy b) -> Poly.equal a b)
+       stack_frames b.stack_frames
+
+let equal f = Result.equal f error_equal
+
+let error_hash { exn; stack_frames } =
+  Hashtbl.hash (Exn.hash exn, List.map stack_frames ~f:Lazy.force)
```

Comment:
Small nitpick: I think we generally prefer `Poly.hash`.

Reply:
Sounds good, changed